flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkRelBuilder.scala

```diff
@@ -167,19 +165,9 @@ object FlinkRelBuilder {
 
   def proto(context: Context): RelBuilderFactory = new RelBuilderFactory() {
     def create(cluster: RelOptCluster, schema: RelOptSchema): RelBuilder = {
+      val clusterContext = cluster.getPlanner.getContext.unwrap(classOf[FlinkContext])
+      val mergedContext = Contexts.chain(clusterContext, context)
```

Comment:
Can we need to chain the contexts here ? Have a quick look at the `ChainContext#unwrap`, it seems that the `Contexts.chain(clusterContext, context)` always returns `clusterContext ` if it is not null, this logic is just inconsistent with the original, because the old code use `context` to unwrap.

Reply:
I think the current behavior is correct whereas the old one was rather wrong or not necessarily wrong, but was throwing away the benefits of having the `unwrap` method.

What do you mean that it returns always `clusterContext`? It always returns `ChainContext` with sub contexts as provided, than you can access particular context through `unwrap` method. The `Contexts.chain` method is exactly for the purpose of merging multiple different contexts.

Let's have a look at an example:
```
val calciteConfig = Contexts.of(
      FLINK_LOGICAL_PROJECT_FACTORY, 
      ...
      )
val flinkConfig = new FlinkConfig(...)

val chainedConfig = Contexts.chain(calciteConfig, flinkConfig)

chainedConfig.unwrap(classOf[FlinkConfig]) // returns flinkConfig
chainedConfig.unwrap(classOf[ProjectFactory]) // returns FLINK_LOGICAL_PROJECT_FACTORY
```

Of course we can talk about which of the contexts should take precedence. You are probably right(though not 100% sure as the merging seems a bit odd to me) that the `context` should be put first. Will change that.

Reply:
The `ChainContext#unwap` is easy to use but a little hard to understand, just like you said, the context returned is very relevant with the sequence of the contexts(for same class contexts), which is not that intuitive. I'm not against to use `ChainContext`, we just need to be caution with the usage.